PSReadLine/PlatformWindows.cs

```diff
@@ -1015,4 +1015,125 @@ private static void TerminateStragglers()
+    ///     owner as well. We verify that by going up the process tree and if we find the foreground process owner among
+    ///     the current process' parent hierarchy, then we consider we found it.
+    /// </para>
+    /// <para>
+    ///     The terminal process is not always the direct parent of the current process, but may be higher in the
+    ///     process tree in case PowerShell is a child of some other console process.
+    /// </para>
+    /// <para>
+    ///     The result is cached for performance reasons. To avoid thread id reuse, we keep the thread handle as well.
+    /// </para>
+    /// </remarks>
+    private static uint? GetTerminalOwnerThreadId()
+    {
+        lock (_terminalOwnerThreadHandleLock)
+        {
+            if (_terminalOwnerThreadHandle.HasValue)
+            {
+                // Already cached, validate the cache and return.
+                uint tid = GetThreadId(_terminalOwnerThreadHandle.Value);
+                if (tid == 0)
+                {
+                    // Failed to get thread handle, let's free the resources.
+                    CloseHandle(_terminalOwnerThreadHandle.Value);
+                    _terminalOwnerThreadHandle = null;
+                    return null;
+                }
+
+                return tid;
+            }
+
+            IntPtr foregroundWindow = GetForegroundWindow();
+            if (foregroundWindow != IntPtr.Zero)
+            {
+                uint tid = GetWindowThreadProcessId(foregroundWindow, out var terminalOwnerPid);
+                if (tid == 0)
+                {
+                    return null;
+                }
+
+                // We've got the process, but is it a correct process? Verify using the process parent hierarchy.
+                // Define a limit not get stuck in case processed form a loop (possible in case pid reuse).
+                const int iterationLimit = 20;
+
+                PROCESS_BASIC_INFORMATION pbi = new();
+                Process process = Process.GetCurrentProcess();
+                int activeProcessId = process.Id;
+                for (int i = 0; i < iterationLimit; ++i)
+                {
+                    int status = NtQueryInformationProcess(process.Handle, 0, out pbi, Marshal.SizeOf(pbi), out _);
+                    activeProcessId = pbi.InheritedFromUniqueProcessId.ToInt32();
+                    if (status != 0 || activeProcessId == 0 || activeProcessId == terminalOwnerPid)
+                        break;
+
+                    try
+                    {
+                        process = Process.GetProcessById(activeProcessId);
+                    }
+                    catch (Exception)
+                    {
+                        // No access to the process, or the process is already dead. Either way, we cannot determine its
+                        // keyboard layout.
+                        return null;
```

Comment:
When this happens, a parent of the currently process was terminated, and thus we will never be able to verify/confirm that the foreground window is actually the one used by the current process. We can still cache the finding in this case (tid = 0), so the layout will always be retrieve by `GetKeyboardLayout(0)` essentially.

Reply:
Also, I think we should retrieve and cache the `tid` in the `OneTimeInit` method. `OneTimeInit` gets called very early (at the pwsh.exe process start in the most common scenario), so it's less likely for a parent process to be terminated at that point.